lib/services/nativescript-migration-service.ts

```diff
@@ -85,92 +117,185 @@ export class NativeScriptMigrationService implements IFrameworkMigrationService
+					this.nativeScriptMigrationConfiguration.tnsModulesProjectPath,
+					this.nativeScriptMigrationConfiguration.tnsModulesBackupName,
+					() => Future.fromResult(),
+					() => this.$fs.unzip(this.getPathToTypingsZip(newVersion),
+					this.nativeScriptMigrationConfiguration.tnsTypingsPath)).wait();
+
+				this.moveAppResourcesDirectory(this.nativeScriptMigrationConfiguration.appResourcesObsoletePath, this.nativeScriptMigrationConfiguration.appResourcesRequiredPath).wait();
+				this.$fs.deleteDirectory(this.nativeScriptMigrationConfiguration.typingsBackupName).wait();
+				this.$fs.deleteDirectory(this.nativeScriptMigrationConfiguration.tnsModulesBackupName).wait();
+			} catch (err) {
+				this.traceError(err);
+				this.restoreOriginalTnsModulesDirectory().wait();
+				this.moveAppResourcesDirectory(this.nativeScriptMigrationConfiguration.appResourcesRequiredPath, this.nativeScriptMigrationConfiguration.appResourcesObsoletePath).wait();
+				this.executeMigrationActions(this.nativeScriptMigrationConfiguration.oldPackageJsonContents,
+					this.nativeScriptMigrationConfiguration.tnsModulesBackupName,
+					this.nativeScriptMigrationConfiguration.tnsModulesProjectPath,
+					() => this.$fs.deleteDirectory(this.nativeScriptMigrationConfiguration.tnsTypingsPath),
+					() => Future.fromResult()).wait();
+
 				this.$errors.failWithoutHelp("Error during migration. Restored original state of the project.");
 			}
 		}).future<void>()();
 	}
 
+	private migrateByGettingTnsModulesFromZip(newVersion: string): IFuture<void> {
+		return (() => {
+			this.promptForConfirmation(`Migrating to ${newVersion} would mean adding ${NativeScriptMigrationService.TNS_MODULES} directory and modifying ${this.$projectConstants.PACKAGE_JSON_NAME} file, are you sure you want to continue?`).wait();
+
+			try {
+				delete this.nativeScriptMigrationConfiguration.packageJsonContents.dependencies[NativeScriptMigrationService.TNS_CORE_MODULES];
+				this.executeMigrationActions(this.nativeScriptMigrationConfiguration.packageJsonContents,
+					this.nativeScriptMigrationConfiguration.tnsTypingsPath,
+					this.nativeScriptMigrationConfiguration.typingsBackupName,
+					() => Future.fromResult(),
+					() => Future.fromResult()).wait();
+
+				this.extractTnsModulesPackage(newVersion, this.nativeScriptMigrationConfiguration.tnsModulesProjectPath).wait();
+				this.$fs.deleteDirectory(this.nativeScriptMigrationConfiguration.typingsBackupName).wait();
+			} catch (err) {
+				this.traceError(err);
+				this.deleteOriginalTnsModulesDirectory().wait();
+				this.executeMigrationActions(this.nativeScriptMigrationConfiguration.oldPackageJsonContents,
+					this.nativeScriptMigrationConfiguration.typingsBackupName,
+					this.nativeScriptMigrationConfiguration.tnsTypingsPath,
+					() => Future.fromResult(),
+					() => Future.fromResult()).wait();
+
+				this.$errors.failWithoutHelp("Error during migration. Restored original state of the project.");
```

Comment:
This message occurs four times because there are four different migrations depending on from and to migration versions. I managed to reuse the rest of the code but the error-throwing code and message I had to repeat sadly.